blocks/theme-editor/contrast-worker.js

```diff
@@ -0,0 +1,244 @@
+  },
+  'button-secondary-background-color': {
+    // shouldContrastWith: ['button-secondary-text-color'],
+  },
+  'button-secondary-text-color-hover': {
+    shouldContrastWith: ['button-secondary-background-color-hover'],
+  },
+  'button-secondary-background-color-hover': {
+    // shouldContrastWith: ['button-secondary-text-color-hover'],
+  },
+};
+
+function evaluateVariable(obj, value) {
+  if (value?.startsWith('var(')) {
+    const varName = value.replace('var(', '').replace(')', '').trim();
+    const newValue = obj[varName].value;
+
+    if (!newValue) {
+      return null;
+    }
+    if (newValue?.startsWith('var(')) {
+      return evaluateVariable(obj, newValue);
+    }
+    return newValue;
+  }
+
+  return value;
+}
+
+function handler(event) {
+  if (event.data.type !== 'contrastCheck') {
+    return;
+  }
+  const cssVars = getCSSVars(event.data.css);
+  const eventCssVars = cssVars.reduce((output, cssVar) => {
+    let [varName, value] = cssVar.split(':');
+    varName = varName.replaceAll('--', '').trim();
+    value = value.replaceAll('--', '').trim();
+
+    // only get values of variables we care about
+    output[varName] = { value };
+
+    return output;
+  }, {});
+
+  const eventVarsKeys = Object.keys(eventCssVars);
+
+  // vars into values
+  for (let index = 0; index < eventVarsKeys.length; index += 1) {
+    const { value } = eventCssVars[eventVarsKeys[index]];
+    const evaluatedValue = evaluateVariable(eventCssVars, value);
+    if (evaluatedValue) {
+      eventCssVars[eventVarsKeys[index]] = { value: evaluatedValue };
+    }
+  }
+
+  // filter out vars we don't care about
+  const filteredCssVars = Object.keys(eventCssVars).reduce((output, key) => {
+    if (cssVariableMap[key]) {
+      output[key] = eventCssVars[key];
+    }
+    return output;
+  }, {});
+
+  const mapKeys = Object.keys(cssVariableMap);
+  const filteredCssVarsKeys = Object.keys(filteredCssVars);
+  if (filteredCssVarsKeys.length !== mapKeys.length) {
+    console.warn('filteredCssVars has missing keys:', mapKeys - filteredCssVarsKeys);
+  }
+
+  const contrastIssues = [];
+
+  for (let index = 0; index < filteredCssVarsKeys.length; index += 1) {
+    const varKey = filteredCssVarsKeys[index];
+
+    const { shouldContrastWith } = cssVariableMap[varKey];
+
+    if (!shouldContrastWith?.length) continue;
+    for (let j = 0; j < shouldContrastWith.length; j += 1) {
+      const requiredThresholds = getRequiredContrastThreshold(varKey, shouldContrastWith[j]);
+      const { ratio, AA, AAA } = calculateContrast(
+        filteredCssVars[varKey].value,
+        filteredCssVars[shouldContrastWith[j]].value,
+        requiredThresholds.AA,
+        requiredThresholds.AAA,
+      );
+
+      if (!AA || !AAA) {
+        contrastIssues.push({
+          var1: varKey,
+          var2: shouldContrastWith[j],
+          ratio,
+          AA,
+          AAA,
+        });
+      }
+    }
+  }
+  self.postMessage({
```

Comment:
If you change variable in the UI, how many times is this triggered ?

Reply:
the contrast is checked every oninput. I tried to debounce the editor onchange, but I couldn't get it working to save my life. I couldn't figure out why the editor value was outdated. if I waited any amount of time the editor.getValue() woud return the state before the debounce